library/Freckle/App/Kafka.hs

```diff
@@ -0,0 +1,132 @@
+{-# LANGUAGE NamedFieldPuns #-}
+
+module Freckle.App.Kafka
+  ( BrokerAddress(..)
+  , HasKafkaBrokerAddress (..)
+  , envKafkaBrokerAddress
+
+  , KafkaProducer (..)
+  , HasKafkaProducer (..)
+  , kafkaProducer
+
+  , produceAsync
+
+  ) where
+
+import Freckle.App.Prelude
+
+import Blammo.Logging
+import Control.Lens (Lens', view)
+import Data.Aeson (ToJSON, encode)
+import Data.ByteString (toStrict)
+import qualified Data.List.NonEmpty as NE
+import Data.Pool (Pool)
+import qualified Data.Pool as Pool
+import qualified Data.Text as T
+import qualified Freckle.App.Env as Env
+import qualified Kafka.Producer as Kafka
+import UnliftIO.Async (async)
+import UnliftIO.Exception (throwString)
+import Yesod.Core.Lens
+import Yesod.Core.Types (HandlerData)
+
+newtype BrokerAddress = BrokerAddress { getBrokerAddress :: Maybe (NonEmpty Kafka.BrokerAddress)}
+
+class HasKafkaBrokerAddress env where
+  kafkaBrokerAddressL :: Lens' env BrokerAddress
+
+instance HasKafkaBrokerAddress site => HasKafkaBrokerAddress (HandlerData child site) where
+  kafkaBrokerAddressL = envL . siteL . kafkaBrokerAddressL
+
+envKafkaBrokerAddress
+  :: Env.Parser Env.Error BrokerAddress
+envKafkaBrokerAddress = BrokerAddress <$> addressOrNothing
+
+ where
+  addressOrNothing = (parseKafkaBrokerAddress <$> parseKey) <|> pure Nothing
+  parseKey = Env.var Env.nonempty "KAFKA_BROKER_ADDRESS" mempty
+  parseKafkaBrokerAddress = NE.nonEmpty . fmap Kafka.BrokerAddress . T.splitOn ","
+
+data KafkaProducer
+  = NullKafkaProducer
+  | KafkaProducerPool (Pool Kafka.KafkaProducer)
+
+class HasKafkaProducer env where
+  kafkaProducerL :: Lens' env KafkaProducer
+
+instance HasKafkaProducer site => HasKafkaProducer (HandlerData child site) where
+  kafkaProducerL = envL . siteL . kafkaProducerL
+
+kafkaProducer
+  :: ( MonadReader env m
+     , HasKafkaBrokerAddress env
+     , MonadUnliftIO m
+     )
+  => Int
+  -- ^ The number of stripes (distinct sub-pools) to maintain.
+  -- The smallest acceptable value is 1.
+  -> NominalDiffTime
+  -- ^ Amount of time for which an unused resource is kept open.
+  -- The smallest acceptable value is 0.5 seconds.
+  --
+  -- The elapsed time before destroying a resource may be a little
+  -- longer than requested, as the reaper thread wakes at 1-second
+  -- intervals.
+  -> Int
+  -- ^ Maximum number of resources to keep open per stripe.  The
+  -- smallest acceptable value is 1.
+  --
+  -- Requests for resources will block if this limit is reached on a
+  -- single stripe, even if other stripes have idle resources
+  -- available.
+  -> m KafkaProducer
+kafkaProducer numStripes idleTime maxResources =
```

Comment:
I struggled with this and had `createKafkaProducerPool` initially. I shied away due to the return type being just `m KafkaProducer`. I wonder if that should be `KafkaProducerPool` too?

Reply:
Oh that is tricky. The persistent functions return `Pool SqlBackend`.

Personally, I would _try_ with it returning `Pool Kafka.KafkaProducer`. This would require refactoring some things that also appeal to me...

Have `envKafkaBrokerAddresses` return `NonEmpty BrokerAddress`, then the user is responsible to use `optional` (or equivalent) and get themselves a `settingsKafkaBrokerAddresses :: Maybe (NonEmpty BrokerAddress)`.

When users initialize things, they'd be responsible to do something like:

```hs
producer <- maybe
  (pure NullKafkaProducer)
  (fmap . KafkaProducerPool createKafkaProducerPool)
  settingsKafkaBrokerAddresses
```

I might also switch your top-level type to:

```hs
data KafkaProducerPool
  = KafkaProducerPool (Pool KafkaProducer)
  | NullKafkaProducerPool
```

Which means less (no?) `qualified` business.

This drops some boilerplate and indirection out of the module itself, without imposing too much on the client, IMO. As I said, I would just try this direction and see. It could be a dead-end for reasons I'm not seeing from here.

Reply:
I was going back and forth about how much to make the client do here, I think the design you have here probably strikes a good balance and seems to work nicely in `fancy-api`